examples: add write+sync/read bandwidth output from global timers

adammoody opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This adds a line to print global write+sync bandwidth and another line for read bandwidth.  When writing to a disk, like the SSD, the write bandwidth numbers can be inflated because the OS is caching data in memory.  Including the sync cost ensures data is on disk.

This is work in progress.  Associated test cases will need to be updated.

### Description
<!--- Describe your changes in detail -->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the UnifyFS code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted.

MichaelBrim commented:
We should revisit what we print for "effective read/write bandwidth" now that we have the `elapsed_sec_all` functionality in the timers. Currently, `effective_BW = total_bytes / max_time`. But that has the limitation that `max_time` may be less than `(max end timestamp - min begin timestamp)`

adammoody commented:
> We should revisit what we print for "effective read/write bandwidth" now that we have the `elapsed_sec_all` functionality in the timers. Currently, `effective_BW = total_bytes / max_time`. But that has the limitation that `max_time` may be less than `(max end timestamp - min begin timestamp)`

@MichaelBrim , what would you like to do with the effective bandwidth lines?

adammoody commented:
@CamStan , I think this will require us to update some of the tests that check output line count.  Can you point me to those files?

CamStan commented:
@adammoody for the `writeread` example, the line count that is tested is lines 71 and 89 here ([t/ci/100-writeread-tests.sh](https://github.com/LLNL/UnifyFS/blob/9144d777f8dd9b1ea17aa1a51978a2f9864c75a6/t/ci/100-writeread-tests.sh#L71)).

If you end up also editing the `write` and `read` examples; the lines are 71 and 89 in [t/ci/110-write-tests.sh](https://github.com/LLNL/UnifyFS/blob/9144d777f8dd9b1ea17aa1a51978a2f9864c75a6/t/ci/110-write-tests.sh#L71) and 71 and 88 in [t/ci/120-read-tests.sh](https://github.com/LLNL/UnifyFS/blob/9144d777f8dd9b1ea17aa1a51978a2f9864c75a6/t/ci/120-read-tests.sh#L71) respectively.

I'll add better comments to these when I revisit them to add more meaningful tests.

MichaelBrim commented:
@adammoody Taking the max value of `elapsed_sec_all` should be the closest approximation of "max end timestamp on any rank" - "min begin timestamp on any rank". With that, I was thinking that a better effective bandwidth calculation would be:
```
max_write_time = test_reduce_double_max(cfg, time_wr.elapsed_sec_all);
eff_write_bw = bandwidth_mib(total_bytes, max_write_time);
```

adammoody commented:
@CamStan and @MichaelBrim , can you take a fresh look at this?

I updated the writeread test to adjust for the new line count, though I haven't tested the change to the test.  I updated effective write/read bandwidth to use the max of the global timers and then dropped the "Global BW" lines that I had added, since the effective BW is basically the same thing now.

adammoody commented:
Example output (13 lines):
```
+ jsrun -r 32 libexec/writeread-static -p n1 -n 4 -c 1048576 -b 16777216 -a 0 -m /unifyfs -f write-static_pn1_n32_c1MB_b16MB.app
File Create Time is 0.004099 s
Aggregate Write BW is 94694.421 MiB/s
Effective Write BW is 89225.809 MiB/s
Minimum Sync Time is 0.000941 s
Maximum Sync Time is 0.017565 s
Global Sync Time is 0.017567 s
Global Write BW with Sync is 50544.189 MiB/s
Stat Time Pre-Laminate is 0.002402 s
Stat Time Pre-Laminate2 is 0.002402 s
File Laminate Time is 0.006329 s
Stat Time Post-Laminate is 0.000095 s
Aggregate Read BW is 3836.929 MiB/s
Effective Read BW is 1359.522 MiB/s
```

MichaelBrim commented on examples/src/writeread.c:
```diff
@@ -317,34 +341,50 @@ int main(int argc, char* argv[])
         eff_read_bw = bandwidth_mib(total_bytes, max_read_time);
 
-        printf("File create time is %.3lf s\n",
+        double write_sync_bw = bandwidth_mib(total_bytes,
```
@adammoody should this use `max_write_time` and `max_sync_time` instead of the local timer values?

adammoody replied:
Good question.  For sync time, the writeread test is now reporting:
```
Min Sync Time = allreduce(MIN, time_sync.elapsed_sec)
Max Sync Time = allreduce(MAX, time_sync.elapsed_sec)
Global Sync Time = time_sync.elapsed_sec_all value from rank 0
```
It's useful to keep all three of those, because the min/max shows the spread of local sync cost across processes, which can be significant, and the global cost shows the cost from before the first process starts until the last process ends.

I could update the max_sync_time to do an allreduce of the elapsed_sec_all value instead, and add new min_local_sync_time and max_local_sync_time variables to capture those.

adammoody commented:
@MichaelBrim , I added more timers/output to try to account for all of these variations:
```
File Create Time is 0.004241 s
Minimum Local Write BW is 2846.342 MiB/s
Maximum Local Write BW is 4633.316 MiB/s
Aggregate Local Write BW is 95808.322 MiB/s
Global Write BW is 91066.744 MiB/s
Minimum Sync Time is 0.000952 s
Maximum Sync Time is 0.018367 s
Global Sync Time is 0.018371 s
Global Write+Sync BW is 50122.369 MiB/s
Stat Time Pre-Laminate is 0.002715 s
Stat Time Pre-Laminate2 is 0.002715 s
File Laminate Time is 0.006732 s
Stat Time Post-Laminate is 0.000199 s
Minimum Local Read BW is 42.581 MiB/s
Maximum Local Read BW is 386.543 MiB/s
Aggregate Local Read BW is 3877.195 MiB/s
Global Read BW is 1362.581 MiB/s
```

MichaelBrim commented:
@adammoody looks good to me.

adammoody commented:
Thanks, @MichaelBrim .
